#                                     Heart Disease Prediction using SVM Dataset

**Problem Description**

Predict the Heart Disease based on given attributes:
* 0 - NO HEART DISEASE
* 1 - HEART DISEASE

**Attribute Information:**
> 1. Age 
> 2. Sex (1 = male; 0 = female) 
> 3. Chest pain type (4 values): 
Value 1: typical angina 
Value 2: atypical angina 
Value 3: non-anginal pain 
Value 4: asymptomatic 
> 4. Resting blood pressure 
> 5. Serum cholestoral in mg/dl 
> 6. Fasting blood sugar > 120 mg/dl
> 7. Resting electrocardiographic results (values 0,1,2): 
-- Value 0: normal 
-- Value 1: having ST-T wave abnormality (T wave inversions and/or ST elevation or depression of > 0.05 mV) 
-- Value 2: showing probable or definite left ventricular hypertrophy by Estes' criteria 
> 8. Maximum heart rate achieved 
> 9. Exercise induced angina 
> 10. Oldpeak = ST depression induced by exercise relative to rest 
> 11. The slope of the peak exercise ST segment 
> 12. Number of major vessels (0-3) colored by flourosopy 
> 13. Thalessimia: 3 = normal; 6 = fixed defect; 7 = reversable defect

In [1]:
    # Loading Python Modules
import numpy as np 
import pandas as pd 
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn import svm 
from sklearn import metrics #Import scikit-learn metrics module for accuracy calculation
from sklearn.metrics import confusion_matrix

In [2]:
    #Loading the dataset
data = pd.read_csv("heart.csv")

In [3]:
data.head()

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   63    1   3       145   233    1        0      150      0      2.3      0   
1   37    1   2       130   250    0        1      187      0      3.5      0   
2   41    0   1       130   204    0        0      172      0      1.4      2   
3   56    1   1       120   236    0        1      178      0      0.8      2   
4   57    0   0       120   354    0        1      163      1      0.6      2   

   ca  thal  target  
0   0     1       1  
1   0     2       1  
2   0     2       1  
3   0     2       1  
4   0     2       1

In [4]:
    #Display basic statistics of data
data.describe()

age         sex          cp    trestbps        chol         fbs  \
count  303.000000  303.000000  303.000000  303.000000  303.000000  303.000000   
mean    54.366337    0.683168    0.966997  131.623762  246.264026    0.148515   
std      9.082101    0.466011    1.032052   17.538143   51.830751    0.356198   
min     29.000000    0.000000    0.000000   94.000000  126.000000    0.000000   
25%     47.500000    0.000000    0.000000  120.000000  211.000000    0.000000   
50%     55.000000    1.000000    1.000000  130.000000  240.000000    0.000000   
75%     61.000000    1.000000    2.000000  140.000000  274.500000    0.000000   
max     77.000000    1.000000    3.000000  200.000000  564.000000    1.000000   

          restecg     thalach       exang     oldpeak       slope          ca  \
count  303.000000  303.000000  303.000000  303.000000  303.000000  303.000000   
mean     0.528053  149.646865    0.326733    1.039604    1.399340    0.729373   
std      0.525860   22.905161    0.469794    1.161075    0.616226    1.022606   
min      0.000000   71.000000    0.000000    0.000000    0.000000    0.000000   
25%      0.000000  133.500000    0.000000    0.000000    1.000000    0.000000   
50%      1.000000  153.000000    0.000000    0.800000    1.000000    0.000000   
75%      1.000000  166.000000    1.000000    1.600000    2.000000    1.000000   
max      2.000000  202.000000    1.000000    6.200000    2.000000    4.000000   

             thal      target  
count  303.000000  303.000000  
mean     2.313531    0.544554  
std      0.612277    0.498835  
min      0.000000    0.000000  
25%      2.000000    0.000000  
50%      2.000000    1.000000  
75%      3.000000    1.000000  
max      3.000000    1.000000

In [5]:
countNoDisease = len(data[data.target == 0])
countHaveDisease = len(data[data.target == 1])
print("Total no. of Patients:", len(data))
print("No. of Patients Having Heart Disease: ", countHaveDisease)
print("No. of Patients Not Having Heart Disease: ", countNoDisease)
print()
print("Percentage of Patients Having Heart Disease: {:.2f}%".format((countHaveDisease / (len(data.target))*100)))
print("Percentage of Patients Not Having Heart Disease: {:.2f}%".format((countNoDisease / (len(data.target))*100)))

Total no. of Patients: 303
No. of Patients Having Heart Disease:  165
No. of Patients Not Having Heart Disease:  138

Percentage of Patients Having Heart Disease: 54.46%
Percentage of Patients Not Having Heart Disease: 45.54%


In [6]:
    #Separate Feature and Target Matrix
features = data.drop('target',axis = 1) 
target = data.target

In [7]:
    #Split dataset into training set and test set
    #70% training and 30% test
x_train, x_test, y_train, y_test = train_test_split(features, target, test_size=0.3,random_state=109) 

**Training and Testing the Machine Learning Model - SVM (Support Vector Machines)**

In [8]:
    #Create a svm Classifier
    # Linear Kernel
model = svm.SVC(kernel='linear') 

    #Train the model using the training sets
model.fit(x_train, y_train)

    #Predict the response for test dataset
y_pred = model.predict(x_test)

In [9]:
    #Model Accuracy
model.score(x_test,y_test)

0.9010989010989011

    Model Accuracy is 90.1%

    Evaluating Accuracy through Confusion Matrix:
    Find the results [TP   FP

                      FN  TN ]

In [10]:
confusion_matrix(y_test,y_pred)

array([[35,  5],
       [ 4, 47]], dtype=int64)

**COMMENTS:** 
* 35 patients were predicted that they **will** have Heart Disease,the Prediction was CORRECT (True-Positive)
* 47 patients were predicted that they **will NOT** have Heart Disease,the Prediction was CORRECT (True-Negative)
* 5 patients were predicted that they **will** have Heart Disease but the Prediction was WRONG (False-Positive)
* 4 patients were predicted that they **will NOT** have Heart Disease but the Prediction was WRONG (False-Negative)

**Comparing with other Machine Learning Model**

In [11]:
from sklearn.linear_model import LinearRegression

model1 = LinearRegression()
model1.fit(x_train, y_train)
y_pred = model1.predict(x_test)
val = model1.score(x_test,y_test)
print("Model Accuracy of Linear Regression is {:0.2f}% ".format(val*100))

Model Accuracy of Linear Regression is 51.05% 


In [12]:
from sklearn.linear_model import LogisticRegression
model2 = LogisticRegression()
model2.fit(x_train, y_train)
y_pred = model2.predict(x_test)
val = model2.score(x_test,y_test)
print("Model Accuracy of Logistic Regression is {:0.2f}% ".format(val*100))

Model Accuracy of Logistic Regression is 91.21% 


D:\Software\Anaconda\lib\site-packages\sklearn\linear_model\_logistic.py:762: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [13]:
from sklearn import tree
model3 = tree.DecisionTreeClassifier()
model3.fit(x_train, y_train)
y_pred = model3.predict(x_test)
val = model3.score(x_test,y_test)
print("Model Accuracy of Decision Tree Classifier is {:0.2f}% ".format(val*100))

Model Accuracy of Decision Tree Classifier is 78.02% 


In [14]:
from sklearn.ensemble import RandomForestClassifier
model4 = RandomForestClassifier(n_estimators=20)
model4.fit(x_train, y_train)
y_pred = model4.predict(x_test)
val = model4.score(x_test,y_test)
print("Model Accuracy of Random Forest Classifier is {:0.2f}% ".format(val*100))

Model Accuracy of Random Forest Classifier is 85.71% 
